core/lib/web3_decl/src/client.rs

```diff
@@ -0,0 +1,456 @@
+//! L2 HTTP client.
+
+use std::{
+    fmt,
+    num::NonZeroUsize,
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+
+use async_trait::async_trait;
+use jsonrpsee::{
+    core::{
+        client::{BatchResponse, ClientT, Error, Subscription, SubscriptionClientT},
+        params::BatchRequestBuilder,
+        traits::ToRpcParams,
+    },
+    http_client::{HttpClient, HttpClientBuilder},
+};
+use serde::de::DeserializeOwned;
+use tokio::time::Instant;
+
+#[derive(Debug, Clone, Copy)]
+enum RateLimitOrigin<'a> {
+    Notification(&'a str),
+    Request(&'a str),
+    BatchRequest(&'a BatchRequestBuilder<'a>),
+    Subscription(&'a str),
+}
+
+impl fmt::Display for RateLimitOrigin<'_> {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Notification(name) => write!(formatter, "notification `{name}`"),
+            Self::Request(name) => write!(formatter, "request `{name}`"),
+            Self::BatchRequest(req) => {
+                let method_names: Vec<_> = req.iter().map(|(name, _)| name).collect();
+                write!(formatter, "batch request {method_names:?}")
+            }
+            Self::Subscription(name) => write!(formatter, "subscription `{name}`"),
+        }
+    }
+}
+
+/// JSON-RPC client for the main node with built-in middleware support.
+///
+/// The client should be used instead of `HttpClient` etc. A single instance of the client should be built
+/// and shared among all tasks run by the node in order to correctly rate-limit requests.
+#[derive(Debug, Clone)]
+pub struct L2Client {
+    inner: HttpClient,
+    rate_limit: SharedRateLimit,
+    component_name: &'static str,
+}
+
+impl L2Client {
+    pub fn builder() -> L2ClientBuilder {
+        L2ClientBuilder::default()
+    }
+
+    /// Sets the component operating this client. This is used in logging etc.
+    pub fn for_component(mut self, component_name: &'static str) -> Self {
+        self.component_name = component_name;
+        self
+    }
+
+    fn report_waiting(&self, origin: RateLimitOrigin, stats: &AcquireStats) {
```

Comment:
Would it make sense to add a counter metric for slow requests (perhaps, with the method / component labels)?

Reply:
Imo yes, or make rate limitting under warn, not debug (or both :P), now most of the partners won't be able to see it and so they won't tell us about it